components/raftstore/src/store/util.rs

```diff
@@ -911,16 +913,22 @@ struct RegionReadProgressCore {
     applied_index: u64,
     // TODO: after region merged, the region's key range is extended and this
     // region wide `safe_ts` should reset to `min(source_safe_ts, target_safe_ts)`
-    safe_ts: u64,
+    read_state: ReadState,
     // `pending_items` is a *sorted* list of `(apply index, safe ts)` item
-    pending_items: VecDeque<(u64, u64)>,
+    pending_items: VecDeque<ReadState>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct ReadState {
```

Comment:
I am confused by `ReadState` with the `RegionReadProgress`. Maybe using a name that can indicate the relation between `idx` and `ts` is better?